src/utils/hooks/useUndoQueue.tsx

```diff
@@ -0,0 +1,55 @@
+import { useState } from 'react'
+
+export const useUndoQueue = <T,>(initialValue?: T) => {
+  const [undoQueue, setUndoQueue] = useState<T[]>([])
+  const [redoQueue, setRedoQueue] = useState<T[]>([])
+  const [currentValue, setCurrentValue] = useState<T | undefined>(initialValue)
+  const [hasChanged, setHasChanged] = useState(false)
+
+  const setValue = (value: T) => {
+    if (value !== undefined) {
+      if (currentValue !== undefined) {
+        setUndoQueue([...undoQueue, currentValue])
+        setRedoQueue([])
+        setHasChanged(true)
+      }
+      setCurrentValue(value)
+    }
+  }
```

Comment:
Probably a bit overkill to create a whole Undo queue mechanism, but it was a fun little task 😄

If we find that it's buggy though, I'm happy to replace it with an external package such as https://www.npmjs.com/package/use-undo